internal/controller/thanosquery_controller.go

```diff
@@ -18,45 +18,142 @@ package controller
 //+kubebuilder:rbac:groups=monitoring.thanos.io,resources=thanosqueries/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=monitoring.thanos.io,resources=thanosqueries/finalizers,verbs=update
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=services;configmaps;serviceaccounts,verbs=get;list;watch;create;update;patch;delete
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the ThanosQuery object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.0/pkg/reconcile
 func (r *ThanosQueryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
 
-	// TODO(user): your logic here
+	query := &monitoringthanosiov1alpha1.ThanosQuery{}
+	err := r.Get(ctx, req.NamespacedName, query)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("thanos query resource not found. ignoring since object may be deleted")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get ThanosQuery")
+		return ctrl.Result{}, err
+	}
+
+	err = r.syncResources(ctx, *query)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *ThanosQueryReconciler) syncResources(ctx context.Context, query monitoringthanosiov1alpha1.ThanosQuery) error {
+	logger := log.FromContext(ctx)
+	var objs []client.Object
+
+	objs = append(objs, r.buildQuerier(query)...)
+
+	var errCount int32
+	for _, obj := range objs {
+		if manifests.IsNamespacedResource(obj) {
+			obj.SetNamespace(query.Namespace)
+			if err := ctrl.SetControllerReference(&query, obj, r.Scheme); err != nil {
+				logger.Error(err, "failed to set controller owner reference to resource")
+				errCount++
+				continue
+			}
+		}
+
+		desired := obj.DeepCopyObject().(client.Object)
+		mutateFn := manifests.MutateFuncFor(obj, desired)
+
+		op, err := ctrl.CreateOrUpdate(ctx, r.Client, obj, mutateFn)
+		if err != nil {
+			logger.Error(
+				err, "failed to create or update resource",
+				"gvk", obj.GetObjectKind().GroupVersionKind().String(),
+				"resource", obj.GetName(),
+				"namespace", obj.GetNamespace(),
+			)
+			errCount++
+			continue
+		}
+
+		logger.V(1).Info(
+			"resource configured",
+			"operation", op, "gvk", obj.GetObjectKind().GroupVersionKind().String(),
+			"resource", obj.GetName(), "namespace", obj.GetNamespace(),
+		)
+	}
+
+	if errCount > 0 {
+		return fmt.Errorf("failed to create or update %d resources for the hashrings", errCount)
```

Comment:
copy pasta typo here with the ref to hashrings - but it was good that we are able to copy and paste :)